### PR TITLE
Add Terraform cloud module registry plugin

### PR DESCRIPTION
**Pull request by nigelwright-roam**
## Hey, I just made a Pull Request!

Added a new plugin to pull terraform modules from Terraform Cloud, and add them to the software catalog. This will also grab all the terraform resources created as part of the module, and add these with dependencies to the module. 
image
image
Related to #8624 
#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [X] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/terraform-cloud-processor-backend/package.json [hunk 1]
```diff
@@ -0,0 +1,43 @@
+{
+  "name": "@backstage/plugin-terraform-cloud-processor-backend",
```
Comment (Rugvip) [inline plugins/terraform-cloud-processor-backend/package.json:2]: I'm thinking this perhaps is better off as a catalog module, similar to the LDAP and MSGraph ones. So it's be something like `plugin-catalog-backend-module-terraform-cloud`

**Review by Rugvip — COMMENTED** [review]
Thank you! 👍 

Just a quick comment before a more thorough review later. You'll also want to skip the `package-lock.json` addition

File: plugins/terraform-cloud-processor-backend/package.json (hunk 1 above)
Comment (nigelwright-roam) [reply plugins/terraform-cloud-processor-backend/package.json:2]: makes sense - will update!

File: plugins/catalog-backend-module-terraform-cloud/src/processors/TerraformCloudProcessor.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+ * limitations under the License.
+ */
+
+import { LocationSpec, Entity } from '@backstage/catalog-model';
+import {
+  CatalogProcessor,
+  CatalogProcessorEmit,
+  results
+} from '@backstage/plugin-catalog-backend';
+import fetch from 'cross-fetch';
```
Comment (Rugvip) [inline plugins/catalog-backend-module-terraform-cloud/src/processors/TerraformCloudProcessor.ts:23]: This can be `node-fetch` instead since it's backend-only

File: plugins/catalog-backend-module-terraform-cloud/src/processors/TerraformCloudProcessor.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    location: LocationSpec,
+    _optional: boolean,
+    emit: CatalogProcessorEmit,
+  ): Promise<boolean> {
+    if (location.type !== 'tf-cloud') {
+      return false;
+    }
+        
+    try {
+             const token = this.config.getOptionalString('integrations.terraform.token');
```
Comment (Rugvip) [inline plugins/catalog-backend-module-terraform-cloud/src/processors/TerraformCloudProcessor.ts:41]: Not sure about this config pattern tbh, I think either it would just be a top-level `terraform.token`, or use the same layout as the other integrations with an array of different terraform integrations that one can select from. I think the plain `terraform.token` would make sense for now, and then we can see where we end up with more terraform integrations going forward. Another slight spin on that could be `terraform.cloud.token`

File: plugins/catalog-backend-module-terraform-cloud/src/processors/TerraformCloudProcessor.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  CatalogProcessor,
+  CatalogProcessorEmit,
+  results
+} from '@backstage/plugin-catalog-backend';
+import fetch from 'cross-fetch';
+import { Config } from '@backstage/config';
+import  { Logger } from 'winston';
+
+export class TfCloudReaderProcessor implements CatalogProcessor {
+    constructor(private readonly config: Config, private readonly logger: Logger) {}
```
Comment (Rugvip) [inline plugins/catalog-backend-module-terraform-cloud/src/processors/TerraformCloudProcessor.ts:28]: When reading config during instantiation we'd typically use the `fromConfig` pattern, see for example https://github.com/backstage/backstage/blob/658131a2900928cf4d020f7f2bedcdf529bb5555/plugins/catalog-backend/src/ingestion/processors/AzureDevOpsDiscoveryProcessor.ts#L43

File: plugins/terraform-cloud-processor-backend/package.json (hunk 1 above)
Comment (Rugvip) [reply plugins/terraform-cloud-processor-backend/package.json:2]: Thanks! seems we have a duplication going on now that needs to be cleaned up :grin:

File: plugins/catalog-backend-module-terraform-cloud/package.json [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  },
+  "dependencies": {
+    "@backstage/backend-common": "^0.9.11",
+    "@backstage/catalog-model": "^0.9.8",
+    "@backstage/config": "^0.1.8",
+    "@backstage/plugin-catalog-backend": "^0.19.2",
+    "@types/express": "*",
+    "cross-fetch": "^3.0.6",
+    "express": "^4.17.1",
+    "express-promise-router": "^4.1.0",
```
Comment (Rugvip) [inline plugins/catalog-backend-module-terraform-cloud/package.json:30]: I think it's worthwhile cleaning up a few of these dependencies that aren't used

File: plugins/catalog-backend-module-terraform-cloud/package.json [hunk 6]
```diff
@@ -0,0 +1,43 @@
+{
+  "name": "@backstage/plugin-catalog-backend-module-terraform-cloud",
+  "version": "0.1.0",
+  "main": "src/index.ts",
+  "types": "src/index.ts",
+  "license": "Apache-2.0",
+  "private": true,
```
Comment (Rugvip) [inline plugins/catalog-backend-module-terraform-cloud/package.json:7]: `false` 😁

**Review by Rugvip — COMMENTED** [review]
Thank you, good stuff! 😁 

Had a more proper look with a couple more thoughts

File: plugins/catalog-backend-module-terraform-cloud/src/processors/TerraformCloudProcessor.ts (hunk 3 above)
Comment (nigelwright-roam) [reply plugins/catalog-backend-module-terraform-cloud/src/processors/TerraformCloudProcessor.ts:41]: Makes sense to me - will go with the `terraform.token` as there really isn't any other auth token used in terraform, so `terraform.cloud.token` might be a bit too wordy :)

**Comment by regicsolutions** [comment]
@nigelwright-roam wondering if this plugin would also work with Terraform Enterprise?

**Comment by nigelwright-roam** [comment]
> wondering if this plugin would also work with Terraform Enterprise?

@regicsolutions  Hmm good question. I believe it should work, but I don't have a Terraform Enterprise instance to test against..cloud/enterprise APIs are generally the same it appears.

File: plugins/catalog-backend-module-terraform-cloud/src/processors/TerraformCloudProcessor.ts (hunk 4 above)
Comment (nigelwright-roam) [reply plugins/catalog-backend-module-terraform-cloud/src/processors/TerraformCloudProcessor.ts]: ok - I've updated to use this pattern, appreciate any and all feedback / pointers on where I've gone wrong here!

**Review by benjdlambert — COMMENTED** [review]
Thanks for this! Just one small nit and I think we're ready to :ship:

**Comment by nigelwright-roam** [comment]
Thanks @benjdlambert - apologies for being blind here what is the nit left to fix? Appreciate the help getting this across the line!

**Comment by benjdlambert** [comment]
@nigelwright-roam hahah I have no idea. 🙈  Something messed up but I can't seem to find what it was. This looks good, just waiting for all the checks to run and should be good to :ship:

**Comment by benjdlambert** [comment]
@nigelwright-roam I think I've fixed the build - was something to do with out of date packages in `package.json` and it not pulling in the latest versions from the workspace I believe. 🤞

**Comment by benjdlambert** [comment]
Hey @nigelwright-roam - I've tried to get this working to get into the release today, but it looks like there's some unit tests failures. Can you take a look and fix these up?

**Comment by nigelwright-roam** [comment]
Hey @benjdlambert - thanks for the assist. I'll take a look shortly and get this sorted.

**Review by Rugvip — COMMENTED** [review]
Hi @nigelwright-roam! Wanna check in whether there's anything we can help out with to get this ready to ship

**Comment by benjdlambert** [comment]
@nigelwright-roam if you want to come pick this up again, please let us know and we can re-open this PR! 👍

**Comment by regicsolutions** [comment]
@nigelwright-roam i tested this with terraform-enterprise and getting a 404 when trying to fetch resources, the target url seems to be attempting to fetch the /1.0.0 version when trying this in postman I am able to replicate the same 404/not-found, wondering if you came across this in terraform-cloud?

**Comment by iiro** [comment]
Hi @nigelwright-roam! :) What would be needed to finish the work...? This would be a really great add-on for any Terraform Cloud/Enterprise user...

**Comment by shatteredeagle** [comment]
Heya @nigelwright-roam, I know this hasn't been looked at in a year but having an integration with TFC would be a killer feature for Backstage! Wanted to echo what @iiro said above and see if there's anything that we can do to help get this pulled in?

**Comment by mcintyre321** [comment]
I'm also really interested in this. What's the status?
